File: python-ref/gvibu_ref/commands/test_cmd.py
"""test / [: evaluate expression."""
import os
import stat
import sys


def _file_exists(path: str) -> bool:
    try:
        os.stat(path)
        return True
    except OSError:
        return False


def _is_file(path: str) -> bool:
    try:
        return stat.S_ISREG(os.stat(path).st_mode)
    except OSError:
        return False


def _is_dir(path: str) -> bool:
    try:
        return stat.S_ISDIR(os.stat(path).st_mode)
    except OSError:
        return False


def _is_executable(path: str) -> bool:
    try:
        st = os.stat(path)
        return bool(st.st_mode & (stat.S_IXUSR | stat.S_IXGRP | stat.S_IXOTH))
    except OSError:
        return False


def _is_writable(path: str) -> bool:
    return os.access(path, os.W_OK)


def _is_readable(path: str) -> bool:
    return os.access(path, os.R_OK)


def _is_nonempty(path: str) -> bool:
    try:
        return os.stat(path).st_size > 0
    except OSError:
        return False


def _eval_unary(op: str, val: str) -> bool:
    if op == '-n':
        return bool(val)
    elif op == '-z':
        return not val
    elif op == '-f':
        return _is_file(val)
    elif op == '-d':
        return _is_dir(val)
    elif op == '-e':
        return _file_exists(val)
    elif op == '-x':
        return _is_executable(val)
    elif op == '-w':
        return _is_writable(val)
    elif op == '-r':
        return _is_readable(val)
    elif op == '-s':
        return _is_nonempty(val)
    return False


def _parse_num(s: str) -> int | None:
    s = s.strip()
    if not s:
        return None
    try:
        return int(s)
    except ValueError:
        return None


def _eval_binary(left: str, op: str, right: str) -> bool:
    if op == '=':
        return left == right
    elif op == '!=':
        return left != right
    elif op == '-eq':
        a, b = _parse_num(left), _parse_num(right)
        return a is not None and b is not None and a == b
    elif op == '-ne':
        a, b = _parse_num(left), _parse_num(right)
        return a is not None and b is not None and a != b
    elif op == '-lt':
        a, b = _parse_num(left), _parse_num(right)
        return a is not None and b is not None and a < b
    elif op == '-le':
        a, b = _parse_num(left), _parse_num(right)
        return a is not None and b is not None and a <= b
    elif op == '-gt':
        a, b = _parse_num(left), _parse_num(right)
        return a is not None and b is not None and a > b
    elif op == '-ge':
        a, b = _parse_num(left), _parse_num(right)
        return a is not None and b is not None and a >= b
    return False


def run(args: list[str]) -> int:
    if not args:
        print("test: missing operand", file=sys.stderr)
        return 1

    test_args: list[str]
    if args[0] == '[':
        if len(args) < 2 or args[-1] != ']':
            print("test: missing ']'", file=sys.stderr)
            return 1
        test_args = args[1:-1]
    else:
        test_args = args

    if not test_args:
        return 1

    # Simple recursive descent evaluator
    # Supported: unary ops, binary ops, ! prefix
    # For simplicity, no -a/-o (AND/OR) support in this version
    i = 0
    not_flag = False

    # Check for leading !
    if test_args[0] == '!':
        not_flag = True
        i = 1

    if i >= len(test_args):
        return 1

    result = False

    # Check for unary op
    if test_args[i] in ('-n', '-z', '-f', '-d', '-e', '-x', '-w', '-r', '-s'):
        if i + 1 >= len(test_args):
            print(f"test: missing argument after '{test_args[i]}'", file=sys.stderr)
            return 1
        result = _eval_unary(test_args[i], test_args[i + 1])
        i += 2
    elif i + 2 < len(test_args) and test_args[i + 1] in ('=', '!=', '-eq', '-ne', '-lt', '-le', '-gt', '-ge'):
        result = _eval_binary(test_args[i], test_args[i + 1], test_args[i + 2])
        i += 3
    else:
        # Plain string — test if non-empty
        result = bool(test_args[i])
        i += 1

    if not_flag:
        result = not result

    return 0 if result else 1

File: rust/src/commands/test_cmd.rs
/// test / [: evaluate expression.
/// Supports `test EXPR` and `[ EXPR ]` forms.
use std::fs;
use std::io::Write;

fn is_file(s: &str) -> bool {
    fs::metadata(s).is_ok()
}

fn is_dir(s: &str) -> bool {
    fs::metadata(s).map(|m| m.is_dir()).unwrap_or(false)
}

fn file_exists(s: &str) -> bool {
    fs::metadata(s).is_ok()
}

fn is_executable(s: &str) -> bool {
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        fs::metadata(s)
            .map(|m| m.permissions().mode() & 0o111 != 0)
            .unwrap_or(false)
    }
    #[cfg(not(unix))]
    {
        let _ = s;
        false
    }
}

fn is_writable(s: &str) -> bool {
    fs::metadata(s)
        .map(|m| !m.permissions().readonly())
        .unwrap_or(false)
}

fn is_readable(s: &str) -> bool {
    fs::metadata(s).is_ok()
}

fn parse_num(s: &str) -> Option<i64> {
    if s.is_empty() {
        return None;
    }
    // Allow leading whitespace and optional sign
    let s = s.trim();
    if s.is_empty() {
        return None;
    }
    s.parse::<i64>().ok()
}

enum Token {
    Unary(String, String),
    Binary(String, String, String),
    Not,
}

fn tokenize(args: &[String]) -> Result<Vec<Token>, String> {
    let mut tokens = Vec::new();
    let mut i = 0;
    while i < args.len() {
        let arg = &args[i];
        match arg.as_str() {
            "!" => {
                tokens.push(Token::Not);
                i += 1;
            }
            "-n" | "-z" | "-f" | "-d" | "-e" | "-x" | "-w" | "-r" | "-s" => {
                if i + 1 >= args.len() {
                    return Err(format!("test: missing argument after '{}'", arg));
                }
                tokens.push(Token::Unary(arg.to_string(), args[i + 1].clone()));
                i += 2;
            }
            "=" | "!=" | "-eq" | "-ne" | "-lt" | "-le" | "-gt" | "-ge" => {
                if tokens.is_empty() {
                    return Err(format!("test: missing argument before '{}'", arg));
                }
                // Find the left operand — should be the last string token
                match tokens.pop() {
                    Some(Token::Unary(_, val)) | Some(Token::Binary(val, _, _)) => {
                        if i + 1 >= args.len() {
                            return Err(format!("test: missing argument after '{}'", arg));
                        }
                        let right = args[i + 1].clone();
                        tokens.push(Token::Binary(val, arg.to_string(), right));
                        i += 2;
                    }
                    Some(Token::Not) => {
                        return Err("test: unexpected '!' before binary operator".to_string());
                    }
                    None => {
                        return Err(format!("test: missing argument before '{}'", arg));
                    }
                }
            }
            _ => {
                // Plain string — treat as unary -n
                tokens.push(Token::Unary("-n".to_string(), arg.clone()));
                i += 1;
            }
        }
    }
    Ok(tokens)
}

fn eval_unary(op: &str, val: &str) -> bool {
    match op {
        "-n" => !val.is_empty(),
        "-z" => val.is_empty(),
        "-f" => is_file(val) && !is_dir(val),
        "-d" => is_dir(val),
        "-e" => file_exists(val),
        "-x" => is_executable(val),
        "-w" => is_writable(val),
        "-r" => is_readable(val),
        "-s" => fs::metadata(val).map(|m| m.len() > 0).unwrap_or(false),
        _ => false,
    }
}

fn eval_binary(left: &str, op: &str, right: &str) -> bool {
    match op {
        "=" => left == right,
        "!=" => left != right,
        "-eq" => parse_num(left).zip(parse_num(right)).map_or(false, |(a, b)| a == b),
        "-ne" => parse_num(left).zip(parse_num(right)).map_or(false, |(a, b)| a != b),
        "-lt" => parse_num(left).zip(parse_num(right)).map_or(false, |(a, b)| a < b),
        "-le" => parse_num(left).zip(parse_num(right)).map_or(false, |(a, b)| a <= b),
        "-gt" => parse_num(left).zip(parse_num(right)).map_or(false, |(a, b)| a > b),
        "-ge" => parse_num(left).zip(parse_num(right)).map_or(false, |(a, b)| a >= b),
        _ => false,
    }
}

fn eval_tokens(tokens: &[Token]) -> Result<bool, String> {
    let mut result = true;
    let mut not_next = false;
    let mut i = 0;

    while i < tokens.len() {
        let token = &tokens[i];
        let val = match token {
            Token::Not => {
                not_next = !not_next;
                i += 1;
                continue;
            }
            Token::Unary(op, val) => eval_unary(op, val),
            Token::Binary(left, op, right) => eval_binary(left, op, right),
        };

        let val = if not_next { !val } else { val };
        not_next = false;

        if i == 0 {
            result = val;
        } else {
            result = val; // In the absence of -a/-o, multiple tokens are AND-like
        }
        i += 1;
    }

    Ok(result)
}

pub fn run(_w: &mut dyn Write, args: &[String]) -> i32 {
    if args.is_empty() {
        eprintln!("test: missing operand");
        return 1;
    }

    let test_args: &[String] = if args[0] == "[" {
        // [ EXPR ] form — need closing ]
        if args.len() < 2 || args[args.len() - 1] != "]" {
            eprintln!("test: missing ']'");
            return 1;
        }
        &args[1..args.len() - 1]
    } else {
        args
    };

    if test_args.is_empty() {
        // test with no arguments returns false (exit 1)
        return 1;
    }

    match tokenize(test_args) {
        Ok(tokens) => {
            match eval_tokens(&tokens) {
                Ok(true) => 0,
                Ok(false) => 1,
                Err(e) => {
                    eprintln!("{}", e);
                    return 1;
                }
            }
        }
        Err(e) => {
            eprintln!("{}", e);
            1
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_eval_unary_n() {
        assert!(eval_unary("-n", "hello"));
        assert!(!eval_unary("-n", ""));
    }

    #[test]
    fn test_eval_unary_z() {
        assert!(eval_unary("-z", ""));
        assert!(!eval_unary("-z", "x"));
    }

    #[test]
    fn test_eval_binary_eq() {
        assert!(eval_binary("a", "=", "a"));
        assert!(!eval_binary("a", "=", "b"));
    }

    #[test]
    fn test_eval_binary_neq() {
        assert!(eval_binary("a", "!=", "b"));
        assert!(!eval_binary("a", "!=", "a"));
    }

    #[test]
    fn test_eval_binary_num() {
        assert!(eval_binary("5", "-eq", "5"));
        assert!(eval_binary("3", "-lt", "5"));
        assert!(eval_binary("7", "-gt", "2"));
        assert!(eval_binary("3", "-le", "3"));
        assert!(eval_binary("5", "-ge", "5"));
        assert!(!eval_binary("5", "-eq", "6"));
    }

    #[test]
    fn test_tokenize_unary() {
        let args = vec!["-n".to_string(), "hello".to_string()];
        let tokens = tokenize(&args).unwrap();
        assert_eq!(tokens.len(), 1);
        match &tokens[0] {
            Token::Unary(op, val) => {
                assert_eq!(op.as_str(), "-n");
                assert_eq!(val, "hello");
            }
            _ => panic!("expected Unary"),
        }
    }

    #[test]
    fn test_tokenize_binary() {
        let args = vec!["a".to_string(), "=".to_string(), "b".to_string()];
        let tokens = tokenize(&args).unwrap();
        assert_eq!(tokens.len(), 1);
        match &tokens[0] {
            Token::Binary(left, op, right) => {
                assert_eq!(left, "a");
                assert_eq!(op.as_str(), "=");
                assert_eq!(right, "b");
            }
            _ => panic!("expected Binary"),
        }
    }

    #[test]
    fn test_test_no_args() {
        assert_eq!(run(&mut std::io::sink(), &[]), 1);
    }

    #[test]
    fn test_test_string() {
        assert_eq!(run(&mut std::io::sink(), &["hello".into()]), 0);
    }

    #[test]
    fn test_test_empty_string() {
        assert_eq!(run(&mut std::io::sink(), &["''".into()]), 0);
    }

    #[test]
    fn test_test_n_flag() {
        assert_eq!(run(&mut std::io::sink(), &["-n".into(), "hello".into()]), 0);
    }

    #[test]
    fn test_test_z_flag() {
        assert_eq!(run(&mut std::io::sink(), &["-z".into(), "".into()]), 0);
    }

    #[test]
    fn test_test_eq() {
        assert_eq!(run(&mut std::io::sink(), &["a".into(), "=".into(), "a".into()]), 0);
        assert_eq!(run(&mut std::io::sink(), &["a".into(), "=".into(), "b".into()]), 1);
    }

    #[test]
    fn test_test_neq() {
        assert_eq!(run(&mut std::io::sink(), &["a".into(), "!=".into(), "b".into()]), 0);
    }

    #[test]
    fn test_test_num_eq() {
        assert_eq!(run(&mut std::io::sink(), &["5".into(), "-eq".into(), "5".into()]), 0);
        assert_eq!(run(&mut std::io::sink(), &["5".into(), "-eq".into(), "6".into()]), 1);
    }

    #[test]
    fn test_test_num_lt() {
        assert_eq!(run(&mut std::io::sink(), &["3".into(), "-lt".into(), "5".into()]), 0);
        assert_eq!(run(&mut std::io::sink(), &["5".into(), "-lt".into(), "3".into()]), 1);
    }

    #[test]
    fn test_test_not() {
        assert_eq!(run(&mut std::io::sink(), &["!".into(), "".into()]), 0);
    }

    #[test]
    fn test_test_bracket_form() {
        assert_eq!(run(&mut std::io::sink(), &["[".into(), "hello".into(), "]".into()]), 0);
    }

    #[test]
    fn test_test_bracket_form_false() {
        assert_eq!(run(&mut std::io::sink(), &["[".into(), "".into(), "]".into()]), 1);
    }

    #[test]
    fn test_test_bracket_missing_close() {
        assert_eq!(run(&mut std::io::sink(), &["[".into(), "hello".into()]), 1);
    }

    #[test]
    fn test_test_dev_null_exists() {
        assert_eq!(run(&mut std::io::sink(), &["-e".into(), "/dev/null".into()]), 0);
    }

    #[test]
    fn test_test_dev_null_is_file() {
        assert_eq!(run(&mut std::io::sink(), &["-f".into(), "/dev/null".into()]), 0);
    }
}
